[WorkloadInstall] Simplify code workload install command line

mandel-macaque opened the pull request:
I am trying to locate a bug that is hitting me in the xamarin-macios ci. I was finding it hard to follow the code and found some small bugs/issues while looking at it. I have tried to make it simpler for people like me :) 

1. I have removed all the not needed enum boxing that was done via Equals. I cannot think of a reason why that boxing was preferred.
2. Negated a number of || to use && which short-circuits making it a little better in performance but more important, easier to understand.
3. Reduced the number of nesting levels.
4. Removed a number of allocations that were not needed but got added to avoid a NRE in a captured variable, maybe Enumerable.Emty was not available when this was landed?
5. Removed the used or the TaskAwaiter which is meant to be used by the compiler but not by us humans.
6. Enabled nullability in the file, which pointed out a number of small errors.


I believe that the reviews will be easier done commit by commit, each commit makes changes to a single function when possible (nullability is an exception).

mandel-macaque committed:
[WorkloadInstall] Remove unused using.

mandel-macaque committed:
[WorkloadInstall] Simplify ValidateWorkloadIdsInput.

mandel-macaque committed:
[WorkloadInstall] Simplify Execute.

mandel-macaque committed:
[WorkloadInstall] Simplify TryRunGarbageCollection.

mandel-macaque committed:
[WorkloadInstall] Simplify InstallWorkloadsWithInstallRecord.

mandel-macaque committed:
[WorkloadInstall] Review GetPackageDownloadUrlsAsync.

Simplify code where possible. More importantly do not use TaskAwaiter in
the select since the class is intended to only be used by the compiler
as per the documentation: https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.taskawaiter-1.getresult?view=net-6.0#remarks
Instead use Task.WhenAll on a list of tasks.

mandel-macaque committed:
[WorkloadInstall] Negate || to shortcircuit and avoid the Amny call in UseTempManifestsToResolvePacksAsync.

mandel-macaque committed:
[WorkloadInstall] Remove not longer needed usings after refactors.

mandel-macaque committed:
[WorkloadInstall] Remove not needed boxing and extra where.

mandel-macaque committed:
[WorkloadInstall] Enable nullability which made the code a little cleaner.

mandel-macaque commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -369,24 +414,23 @@ private async Task DownloadToOfflineCacheAsync(IEnumerable<WorkloadId> workloadI
-                    installer.DownloadToOfflineCache(workloadId, offlineCache, includePreviews);
-                }
+                case InstallationUnit.Packs:
+                    var packInstaller= _workloadInstaller.GetPackInstaller();
```
Comment to self: nit, missing space

mandel-macaque replied:
```suggestion
                    var packInstaller = _workloadInstaller.GetPackInstaller();
```

premun commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -395,15 +439,14 @@ private async Task DownloadToOfflineCacheAsync(IEnumerable<WorkloadId> workloadI
         }
 
-        private IEnumerable<PackInfo> GetPacksToInstall(IEnumerable<WorkloadId> workloadIds)
+        IEnumerable<PackInfo?> GetPacksToInstall(IEnumerable<WorkloadId> workloadIds)
```
If you ever figure out how to make Roslyn tell this should be non-nullable becase you do `Where(not null)`, let me know.

I know about `Select(x => x!)` and `Cast<>()` but they seem like extra leg work..

mandel-macaque replied:
I know how to, you can use the more SQL looking syntax, I used it here: https://github.com/dotnet/sdk/pull/24028/files#diff-498ad5982ef0f101943528170b9fb47715442c897bf94ef0c8c1f2438f4f8effR361

Not ideal, that's why I do it in the caller and not the callee.

KalleOlaviNiemitalo commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -64,27 +67,41 @@ internal class WorkloadInstallCommand : CommandBase
+            {
+                _tempDirPath = tempDirPath;
+            }
+            
 
             var configOption = parseResult.GetValueForOption(WorkloadInstallCommandParser.ConfigOption);
             var sourceOption = parseResult.GetValueForOption(WorkloadInstallCommandParser.SourceOption);
-            _packageSourceLocation = string.IsNullOrEmpty(configOption) && (sourceOption == null || !sourceOption.Any()) ? null :
-                new PackageSourceLocation(string.IsNullOrEmpty(configOption) ? null : new FilePath(configOption), sourceFeedOverrides: sourceOption);
+            if (string.IsNullOrEmpty(configOption) && sourceOption != null && sourceOption.Any())
+            {
+                _packageSourceLocation = null;
+            }
+            else
+            {
+                _packageSourceLocation = new(string.IsNullOrEmpty(configOption) ? null : new FilePath(configOption), sourceFeedOverrides: sourceOption);
+            }
```
That is not equivalent to the old code.  If `string.IsNullOrEmpty(configOption) && sourceOption == null`, then the old code would assign `_packageSourceLocation = null`, but your new code goes to the other branch instead.

KalleOlaviNiemitalo replied:
If you want to use `&&`, then `string.IsNullOrEmpty(configOption) && !(sourceOption != null && sourceOption.Any())` is equivalent to the old code, but I wouldn't say it's clearer than the old code. I feel there's a nice symmetry between string.IsNull***Or***Empty and `||`.

mandel-macaque replied:
yep, you are correct, I'm missing a ! around the &&, good catch.

KalleOlaviNiemitalo replied:
`||` does short-circuit.  `sourceOption == null || !sourceOption.Any()` does not call `sourceOption.Any()` if `sourceOption == null` is true.

KalleOlaviNiemitalo commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -358,10 +358,10 @@ async Task<IEnumerable<string>> GetPackageDownloadUrlsAsync(IEnumerable<Workload
         }
 
-        private async Task UseTempManifestsToResolvePacksAsync(DirectoryPath tempPath, bool includePreview)
+        async Task UseTempManifestsToResolvePacksAsync(DirectoryPath tempPath, bool includePreview)
```
This goes against coding style <https://github.com/dotnet/runtime/blob/a55d5300457004aff05a603ba9960290517f5681/docs/coding-guidelines/coding-style.md?plain=1#L14-L16>

KalleOlaviNiemitalo commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -358,10 +358,10 @@ async Task<IEnumerable<string>> GetPackageDownloadUrlsAsync(IEnumerable<Workload
             }
         }
 
-        private async Task UseTempManifestsToResolvePacksAsync(DirectoryPath tempPath, bool includePreview)
+        async Task UseTempManifestsToResolvePacksAsync(DirectoryPath tempPath, bool includePreview)
         {
             var manifestPackagePaths = await _workloadManifestUpdater.DownloadManifestPackagesAsync(includePreview, tempPath);
-            if (manifestPackagePaths == null || !manifestPackagePaths.Any())
+            if (manifestPackagePaths != null && manifestPackagePaths.Any())
             {
                 _reporter.WriteLine(LocalizableStrings.SkippingManifestUpdate);
```
This changes the behavior to the opposite.

mandel-macaque replied:
Sorry, i'm used to the mono coding style.

mandel-macaque replied:
Good catch (should not doo boolean logic at 3 am :( )

KalleOlaviNiemitalo commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -99,22 +99,19 @@ internal class WorkloadInstallCommand : CommandBase
-                        throw new GracefulException(string.Format(LocalizableStrings.WorkloadNotRecognized, workloadId), isUserError: false);
-                    }
-                }
+                if (availableWorkloads.Any(workload => workload.Id.ToString() == workloadId))
+                    continue;
```
This allocates an extra closure and a non-static delegate, but perhaps fewer enumerators. I don't know whether efficiency is important here.

mandel-macaque replied:
Should be doing a lot less enumerations because Any will break ASAP.

mandel-macaque commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -69,27 +67,41 @@ internal class WorkloadInstallCommand : CommandBase
             var sourceOption = parseResult.GetValueForOption(WorkloadInstallCommandParser.SourceOption);
-            _packageSourceLocation = string.IsNullOrEmpty(configOption) && (sourceOption == null || !sourceOption.Any()) ? null :
-                new PackageSourceLocation(string.IsNullOrEmpty(configOption) ? null : new FilePath(configOption), sourceFeedOverrides: sourceOption);
+            if (string.IsNullOrEmpty(configOption) && sourceOption != null && sourceOption.Any())
```
```suggestion
            if (string.IsNullOrEmpty(configOption) && !(sourceOption != null && sourceOption.Any()))
```

mandel-macaque commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -305,39 +345,44 @@ private async Task<IEnumerable<string>> GetPackageDownloadUrlsAsync(IEnumerable<
         }
 
-        private async Task UseTempManifestsToResolvePacksAsync(DirectoryPath tempPath, bool includePreview)
+        async Task UseTempManifestsToResolvePacksAsync(DirectoryPath tempPath, bool includePreview)
```
```suggestion
        private async Task UseTempManifestsToResolvePacksAsync(DirectoryPath tempPath, bool includePreview)
```

mandel-macaque commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -305,39 +345,44 @@ private async Task<IEnumerable<string>> GetPackageDownloadUrlsAsync(IEnumerable<
         {
             var manifestPackagePaths = await _workloadManifestUpdater.DownloadManifestPackagesAsync(includePreview, tempPath);
-            if (manifestPackagePaths == null || !manifestPackagePaths.Any())
+            if (manifestPackagePaths != null && manifestPackagePaths.Any())
```
```suggestion
            if (!(manifestPackagePaths != null && manifestPackagePaths.Any()))
```

mandel-macaque commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -395,15 +439,14 @@ private async Task DownloadToOfflineCacheAsync(IEnumerable<WorkloadId> workloadI
         }
 
-        private IEnumerable<PackInfo> GetPacksToInstall(IEnumerable<WorkloadId> workloadIds)
+        IEnumerable<PackInfo?> GetPacksToInstall(IEnumerable<WorkloadId> workloadIds)
```
```suggestion
        private IEnumerable<PackInfo?> GetPacksToInstall(IEnumerable<WorkloadId> workloadIds)
```

mandel-macaque commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -347,9 +392,9 @@ private async Task UseTempManifestsToResolvePacksAsync(DirectoryPath tempPath, b
         }
 
-        private async Task DownloadToOfflineCacheAsync(IEnumerable<WorkloadId> workloadIds, DirectoryPath offlineCache, bool skipManifestUpdate, bool includePreviews)
+        async Task DownloadToOfflineCacheAsync(IEnumerable<WorkloadId> workloadIds, DirectoryPath offlineCache, bool skipManifestUpdate, bool includePreviews)
```
```suggestion
        private async Task DownloadToOfflineCacheAsync(IEnumerable<WorkloadId> workloadIds, DirectoryPath offlineCache, bool skipManifestUpdate, bool includePreviews)
```

mandel-macaque commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -198,99 +223,114 @@ public void InstallWorkloads(IEnumerable<WorkloadId> workloadIds, bool skipManif
         }
 
-        private async Task<IEnumerable<string>> GetPackageDownloadUrlsAsync(IEnumerable<WorkloadId> workloadIds, bool skipManifestUpdate, bool includePreview)
+        async Task<IEnumerable<string>> GetPackageDownloadUrlsAsync(IEnumerable<WorkloadId> workloadIds, bool skipManifestUpdate, bool includePreview)
```
```suggestion
        private async Task<IEnumerable<string>> GetPackageDownloadUrlsAsync(IEnumerable<WorkloadId> workloadIds, bool skipManifestUpdate, bool includePreview)
```

mandel-macaque commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -198,99 +223,114 @@ public void InstallWorkloads(IEnumerable<WorkloadId> workloadIds, bool skipManif
         }
 
-        private void InstallWorkloadsWithInstallRecord(
+        void InstallPacksWithInstallRecord(
```
```suggestion
        private void InstallPacksWithInstallRecord(
```

mandel-macaque commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -102,64 +114,77 @@ internal class WorkloadInstallCommand : CommandBase
+            return _workloadInstaller.ExitCode;
+        }
+
+        int DownloadToCache()
```
```suggestion
        private int DownloadToCache()
```

mandel-macaque commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -102,64 +114,77 @@ internal class WorkloadInstallCommand : CommandBase
+            }
+        }
+
+        int PrintDownloadLink() { 
```
```suggestion
        private int PrintDownloadLink() { 
```

mandel-macaque committed:
Address reviews

Fix to cases in which a bad boolean logic change was applied. Re-add private since the NetCore guidelines are the opposite to the mono ones.

mandel-macaque committed:
Update src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs

mandel-macaque committed:
Update src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs

marcpopMSFT commented:
Following up on old PRs. The merge conflicts are now fairly involved. If you want to resolve them we can take a look but otherwise I think we'll close this out. Thanks for the effort.

MiYanni commented:
I looked over the changes and much of this code was extracted into a base class, `InstallingWorkloadCommands`. I didn't see anything that stood out as a fundamental improvement; mostly formatting and general cleanliness.

I'd like to do improvements like these across the entire repo to improve readability/ease-of-use. I plan on adding editorconfig rules so that things remain tidy. Unfortunately, this particular set of changes is out-of-date, so I'm closing the PR. Feel free to reopen if you want some of these changes with the current state of the code.
